File: src/core/primitives.py
import numpy as np
import math
from .mesh import Mesh


def create_cube(size=1.0, segments=1):
    m = Mesh("Cube")
    s = size / 2.0
    faces = [
        ((-s, -s, s), (s, -s, s), (s, s, s), (-s, s, s), (0, 0, 1)),
        ((s, -s, -s), (-s, -s, -s), (-s, s, -s), (s, s, -s), (0, 0, -1)),
        ((-s, -s, -s), (-s, -s, s), (-s, s, s), (-s, s, -s), (-1, 0, 0)),
        ((s, -s, s), (s, -s, -s), (s, s, -s), (s, s, s), (1, 0, 0)),
        ((-s, s, s), (s, s, s), (s, s, -s), (-s, s, -s), (0, 1, 0)),
        ((-s, -s, -s), (s, -s, -s), (s, -s, s), (-s, -s, s), (0, -1, 0)),
    ]
    uvs = [(0, 0), (1, 0), (1, 1), (0, 1)]
    for p0, p1, p2, p3, n in faces:
        base = len(m.vertices)
        for i, (p, uv) in enumerate(zip([p0, p1, p2, p3], uvs)):
            m.add_vertex(p, n, uv)
        m.add_face([base, base + 1, base + 2, base + 3])
    m.compute_normals()
    return m


def create_sphere(radius=0.5, segments=32, rings=16):
    m = Mesh("Sphere")
    for ring in range(rings + 1):
        phi = math.pi * ring / rings
        for seg in range(segments + 1):
            theta = 2.0 * math.pi * seg / segments
            x = radius * math.sin(phi) * math.cos(theta)
            y = radius * math.cos(phi)
            z = radius * math.sin(phi) * math.sin(theta)
            n = np.array([x, y, z], dtype=np.float32)
            norm = np.linalg.norm(n)
            if norm > 1e-8:
                n /= norm
            u = seg / segments
            v = ring / rings
            m.add_vertex([x, y, z], n, [u, v])
    for ring in range(rings):
        for seg in range(segments):
            i0 = ring * (segments + 1) + seg
            i1 = i0 + 1
            i2 = i0 + (segments + 1)
            i3 = i2 + 1
            m.add_face([i0, i2, i3, i1])
    return m


def create_cylinder(radius=0.5, height=1.0, segments=32):
    m = Mesh("Cylinder")
    half_h = height / 2.0
    for i in range(segments + 1):
        theta = 2.0 * math.pi * i / segments
        x = radius * math.cos(theta)
        z = radius * math.sin(theta)
        u = i / segments
        nx = math.cos(theta)
        nz = math.sin(theta)
        m.add_vertex([x, half_h, z], [0, 1, 0], [u, 1])
        m.add_vertex([x, -half_h, z], [0, -1, 0], [u, 0])
        side_n = np.array([nx, 0, nz], dtype=np.float32)
        m.add_vertex([x, half_h, z], side_n, [u, 1])
        m.add_vertex([x, -half_h, z], side_n, [u, 0])
    base_center_top = len(m.vertices)
    m.add_vertex([0, half_h, 0], [0, 1, 0], [0.5, 0.5])
    base_center_bottom = len(m.vertices)
    m.add_vertex([0, -half_h, 0], [0, -1, 0], [0.5, 0.5])
    for i in range(segments):
        s = i * 4
        m.add_face([s, s + 4, s + 6, s + 2])
        m.add_face([s + 1, s + 3, s + 5, s + 7])
        m.add_face([base_center_top, s + 2, s])
        m.add_face([base_center_bottom, s + 1, s + 3])
    return m


def create_cone(radius=0.5, height=1.0, segments=32):
    m = Mesh("Cone")
    half_h = height / 2.0
    tip_idx = m.add_vertex([0, half_h, 0], [0, 1, 0], [0.5, 1.0])
    base_center = m.add_vertex([0, -half_h, 0], [0, -1, 0], [0.5, 0.5])
    for i in range(segments + 1):
        theta = 2.0 * math.pi * i / segments
        x = radius * math.cos(theta)
        z = radius * math.sin(theta)
        nx = math.cos(theta)
        nz = math.sin(theta)
        slope = np.array([nx * height, radius, nz * height], dtype=np.float32)
        norm = np.linalg.norm(slope)
        if norm > 1e-8:
            slope /= norm
        u = i / segments
        m.add_vertex([x, -half_h, z], slope, [u, 0])
    for i in range(segments):
        base = i + 2
        next_base = base + 1
        m.add_face([tip_idx, next_base, base])
        m.add_face([base_center, base, next_base])
    return m


def create_torus(major_radius=0.5, minor_radius=0.2, major_segments=32, minor_segments=16):
    m = Mesh("Torus")
    for i in range(major_segments + 1):
        theta = 2.0 * math.pi * i / major_segments
        cos_theta = math.cos(theta)
        sin_theta = math.sin(theta)
        for j in range(minor_segments + 1):
            phi = 2.0 * math.pi * j / minor_segments
            cos_phi = math.cos(phi)
            sin_phi = math.sin(phi)
            x = (major_radius + minor_radius * cos_phi) * cos_theta
            y = minor_radius * sin_phi
            z = (major_radius + minor_radius * cos_phi) * sin_theta
            nx = cos_phi * cos_theta
            ny = sin_phi
            nz = cos_phi * sin_theta
            u = i / major_segments
            v = j / minor_segments
            m.add_vertex([x, y, z], [nx, ny, nz], [u, v])
    for i in range(major_segments):
        for j in range(minor_segments):
            i0 = i * (minor_segments + 1) + j
            i1 = i0 + 1
            i2 = i0 + (minor_segments + 1)
            i3 = i2 + 1
            m.add_face([i0, i2, i3, i1])
    return m


def create_plane(size=1.0, subdivisions=1):
    m = Mesh("Plane")
    half = size / 2.0
    step = size / subdivisions
    for iz in range(subdivisions + 1):
        for ix in range(subdivisions + 1):
            x = -half + ix * step
            z = -half + iz * step
            u = ix / subdivisions
            v = iz / subdivisions
            m.add_vertex([x, 0, z], [0, 1, 0], [u, v])
    for iz in range(subdivisions):
        for ix in range(subdivisions):
            i0 = iz * (subdivisions + 1) + ix
            i1 = i0 + 1
            i2 = i0 + (subdivisions + 1)
            i3 = i2 + 1
            m.add_face([i0, i2, i3, i1])
    return m


def create_grid(size=20.0, subdivisions=20):
    m = Mesh("Grid")
    half = size / 2.0
    step = size / subdivisions
    for i in range(subdivisions + 1):
        pos = -half + i * step
        idx0 = m.add_vertex([pos, 0, -half], [0, 1, 0], [0, 0], [0.4, 0.4, 0.4, 1.0])
        idx1 = m.add_vertex([pos, 0, half], [0, 1, 0], [0, 1], [0.4, 0.4, 0.4, 1.0])
        m.add_face([idx0, idx1])
        idx0 = m.add_vertex([-half, 0, pos], [0, 1, 0], [0, 0], [0.4, 0.4, 0.4, 1.0])
        idx1 = m.add_vertex([half, 0, pos], [0, 1, 0], [1, 0], [0.4, 0.4, 0.4, 1.0])
        m.add_face([idx0, idx1])
    return m


def create_arrow(length=2.0, head_length=0.6, head_radius=0.2, shaft_radius=0.05, segments=16):
    m = Mesh("Arrow")
    half_shaft = (length - head_length) / 2.0
    shaft = create_cylinder(shaft_radius, length - head_length, segments)
    for v in shaft.vertices:
        v.position[1] += half_shaft
    for f in shaft.faces:
        m.add_face([f_v + len(m.vertices) for f_v in f.vertices])
    m.vertices.extend(shaft.vertices)
    head = create_cone(head_radius, head_length, segments)
    for v in head.vertices:
        v.position[1] += half_shaft + (length - head_length) / 2.0 + head_length / 2.0
    offset = len(m.vertices)
    for f in head.faces:
        m.add_face([f_v + offset for f_v in f.vertices])
    m.vertices.extend(head.vertices)
    m.compute_normals()
    return m


def create_torus_knot(p=2, q=3, major_radius=1.0, tube_radius=0.15, segments=128, tube_segments=16):
    m = Mesh("TorusKnot")
    points = []
    for i in range(segments + 1):
        t = 2.0 * math.pi * i / segments
        r = major_radius * (2.0 + math.cos(q * t)) / 3.0
        x = r * math.cos(p * t)
        y = r * math.sin(p * t)
        z = major_radius * math.sin(q * t) / 3.0
        points.append(np.array([x, y, z], dtype=np.float32))
    for i in range(segments):
        t0 = 2.0 * math.pi * i / segments
        t1 = 2.0 * math.pi * (i + 1) / segments
        p0 = points[i]
        p1 = points[i + 1]
        tangent = p1 - p0
        norm = np.linalg.norm(tangent)
        if norm > 1e-8:
            tangent /= norm
        up = np.array([0, 1, 0], dtype=np.float32)
        if abs(np.dot(tangent, up)) > 0.99:
            up = np.array([1, 0, 0], dtype=np.float32)
        side = np.cross(tangent, up)
        side = side / (np.linalg.norm(side) + 1e-8)
        bi = np.cross(tangent, side)
        bi = bi / (np.linalg.norm(bi) + 1e-8)
        for j in range(tube_segments + 1):
            angle = 2.0 * math.pi * j / tube_segments
            cx = tube_radius * math.cos(angle)
            cy = tube_radius * math.sin(angle)
            point = p0 + cx * side + cy * bi
            n = (point - p0)
            n = n / (np.linalg.norm(n) + 1e-8)
            u = i / segments
            v = j / tube_segments
            m.add_vertex(point, n, [u, v])
    for i in range(segments):
        for j in range(tube_segments):
            i0 = i * (tube_segments + 1) + j
            i1 = i0 + 1
            i2 = i0 + (tube_segments + 1)
            i3 = i2 + 1
            m.add_face([i0, i2, i3, i1])
    m.compute_normals()
    return m


def create_ico_sphere(radius=0.5, subdivisions=2):
    m = Mesh("IcoSphere")
    t = (1.0 + math.sqrt(5.0)) / 2.0
    verts = [
        [-1, t, 0], [1, t, 0], [-1, -t, 0], [1, -t, 0],
        [0, -1, t], [0, 1, t], [0, -1, -t], [0, 1, -t],
        [t, 0, -1], [t, 0, 1], [-t, 0, -1], [-t, 0, 1],
    ]
    for v in verts:
        n = np.array(v, dtype=np.float32)
        n = n / (np.linalg.norm(n) + 1e-8)
        p = n * radius
        m.add_vertex(p, n, [0, 0])
    tris = [
        (0, 11, 5), (0, 5, 1), (0, 1, 7), (0, 7, 10), (0, 10, 11),
        (1, 5, 9), (5, 11, 4), (11, 10, 2), (10, 7, 6), (7, 1, 8),
        (3, 9, 4), (3, 4, 2), (3, 2, 6), (3, 6, 8), (3, 8, 9),
        (4, 9, 5), (2, 4, 11), (6, 2, 10), (8, 6, 7), (9, 8, 1),
    ]
    for tri in tris:
        m.add_face(list(tri))
    for _ in range(subdivisions):
        m.subdivide()
        for v in m.vertices:
            n = v.position / (np.linalg.norm(v.position) + 1e-8)
            v.position = n * radius
    m.compute_normals()
    return m


PRIMITIVE_CREATORS = {
    "cube": create_cube,
    "sphere": create_sphere,
    "cylinder": create_cylinder,
    "cone": create_cone,
    "torus": create_torus,
    "plane": create_plane,
    "torus_knot": create_torus_knot,
    "ico_sphere": create_ico_sphere,
}

File: cpp/src/primitives.cpp
#include "primitives.h"
#include <cmath>

#ifndef M_PI
#define M_PI 3.14159265358979323846
#endif

namespace ocp {

Mesh create_cube(float size) {
    Mesh m; m.name = "Cube";
    float s = size * 0.5f;
    vec3 normals[6] = {{0,0,1},{0,0,-1},{-1,0,0},{1,0,0},{0,1,0},{0,-1,0}};
    vec3 verts[6][4] = {
        {{-s,-s, s},{ s,-s, s},{ s, s, s},{-s, s, s}},
        {{ s,-s,-s},{-s,-s,-s},{-s, s,-s},{ s, s,-s}},
        {{-s,-s,-s},{-s,-s, s},{-s, s, s},{-s, s,-s}},
        {{ s,-s, s},{ s,-s,-s},{ s, s,-s},{ s, s, s}},
        {{-s, s, s},{ s, s, s},{ s, s,-s},{-s, s,-s}},
        {{-s,-s,-s},{ s,-s,-s},{ s,-s, s},{-s,-s, s}}
    };
    for (int i = 0; i < 6; i++) {
        uint32_t base = (uint32_t)m.vertices.size();
        for (int j = 0; j < 4; j++) {
            m.add_vertex(verts[i][j], normals[i], vec2(j == 0 || j == 3 ? 0 : 1, j < 2 ? 0 : 1));
        }
        m.add_face({base, base+1, base+2, base+3});
    }
    m.compute_normals();
    return m;
}

Mesh create_sphere(float radius, int segments, int rings) {
    Mesh m; m.name = "Sphere";
    for (int r = 0; r <= rings; r++) {
        float phi = (float)M_PI * r / rings;
        for (int s = 0; s <= segments; s++) {
            float theta = 2.0f * (float)M_PI * s / segments;
            vec3 pos(radius * sinf(phi) * cosf(theta), radius * cosf(phi), radius * sinf(phi) * sinf(theta));
            vec3 nrm = glm::normalize(pos);
            vec2 uv((float)s / segments, (float)r / rings);
            m.add_vertex(pos, nrm, uv);
        }
    }
    for (int r = 0; r < rings; r++) {
        for (int s = 0; s < segments; s++) {
            uint32_t i0 = r * (segments + 1) + s;
            uint32_t i1 = i0 + 1;
            uint32_t i2 = i0 + segments + 1;
            uint32_t i3 = i2 + 1;
            m.add_face({i0, i2, i3, i1});
        }
    }
    m.compute_normals();
    return m;
}

Mesh create_cylinder(float radius, float height, int segments) {
    Mesh m; m.name = "Cylinder";
    float hh = height * 0.5f;

    uint32_t top_center = m.add_vertex(vec3(0, hh, 0), vec3(0, 1, 0), vec2(0.5f, 0.5f));
    uint32_t bot_center = m.add_vertex(vec3(0, -hh, 0), vec3(0, -1, 0), vec2(0.5f, 0.5f));

    uint32_t top_ring = (uint32_t)m.vertices.size();
    uint32_t bot_ring = top_ring + (uint32_t)(segments + 1);
    uint32_t side_top = bot_ring + (uint32_t)(segments + 1);
    uint32_t side_bot = side_top + (uint32_t)(segments + 1);

    for (int i = 0; i <= segments; i++) {
        float theta = 2.0f * (float)M_PI * i / segments;
        float ct = cosf(theta), st = sinf(theta);
        float u = (float)i / segments;
        vec3 radial(ct, 0, st);

        m.add_vertex(vec3(ct * radius, hh, st * radius), vec3(0, 1, 0), vec2(u, 1));
        m.add_vertex(vec3(ct * radius, -hh, st * radius), vec3(0, -1, 0), vec2(u, 0));
        m.add_vertex(vec3(ct * radius, hh, st * radius), radial, vec2(u, 1));
        m.add_vertex(vec3(ct * radius, -hh, st * radius), radial, vec2(u, 0));
    }

    for (int i = 0; i < segments; i++) {
        m.add_face({top_center, top_ring + (uint32_t)i, top_ring + (uint32_t)(i + 1)});
        m.add_face({bot_center, bot_ring + (uint32_t)(i + 1), bot_ring + (uint32_t)i});
        uint32_t st_i = side_top + (uint32_t)i;
        uint32_t st_i1 = side_top + (uint32_t)(i + 1);
        uint32_t sb_i = side_bot + (uint32_t)i;
        uint32_t sb_i1 = side_bot + (uint32_t)(i + 1);
        m.add_face({st_i, st_i1, sb_i1, sb_i});
    }

    m.compute_normals();
    return m;
}

Mesh create_cone(float radius, float height, int segments) {
    Mesh m; m.name = "Cone";
    float hh = height * 0.5f;
    uint32_t tip = m.add_vertex(vec3(0, hh, 0), vec3(0, 1, 0));
    uint32_t center = m.add_vertex(vec3(0, -hh, 0), vec3(0, -1, 0));
    for (int i = 0; i <= segments; i++) {
        float theta = 2.0f * (float)M_PI * i / segments;
        float ct = cosf(theta), st = sinf(theta);
        vec3 pos(ct * radius, -hh, st * radius);
        vec3 slope = glm::normalize(vec3(ct * height, radius, st * height));
        m.add_vertex(pos, slope, vec2((float)i / segments, 0));
    }
    for (int i = 0; i < segments; i++) {
        m.add_face({tip, (uint32_t)(i + 2), (uint32_t)(i + 3)});
        m.add_face({center, (uint32_t)(i + 3), (uint32_t)(i + 2)});
    }
    m.compute_normals();
    return m;
}

Mesh create_torus(float major_r, float minor_r, int major_seg, int minor_seg) {
    Mesh m; m.name = "Torus";
    for (int i = 0; i <= major_seg; i++) {
        float theta = 2.0f * (float)M_PI * i / major_seg;
        float ct = cosf(theta), st = sinf(theta);
        for (int j = 0; j <= minor_seg; j++) {
            float phi = 2.0f * (float)M_PI * j / minor_seg;
            float cp = cosf(phi), sp = sinf(phi);
            float x = (major_r + minor_r * cp) * ct;
            float y = minor_r * sp;
            float z = (major_r + minor_r * cp) * st;
            vec3 nrm = glm::normalize(vec3(cp * ct, sp, cp * st));
            vec2 uv((float)i / major_seg, (float)j / minor_seg);
            m.add_vertex(vec3(x, y, z), nrm, uv);
        }
    }
    for (int i = 0; i < major_seg; i++) {
        for (int j = 0; j < minor_seg; j++) {
            uint32_t a = i * (minor_seg + 1) + j;
            uint32_t b = a + minor_seg + 1;
            m.add_face({a, b, b + 1, a + 1});
        }
    }
    m.compute_normals();
    return m;
}

Mesh create_plane(float size, int subdivisions) {
    Mesh m; m.name = "Plane";
    float hs = size * 0.5f;
    for (int z = 0; z <= subdivisions; z++) {
        for (int x = 0; x <= subdivisions; x++) {
            float fx = (float)x / subdivisions, fz = (float)z / subdivisions;
            m.add_vertex(vec3(-hs + fx * size, 0, -hs + fz * size), vec3(0, 1, 0), vec2(fx, fz));
        }
    }
    for (int z = 0; z < subdivisions; z++) {
        for (int x = 0; x < subdivisions; x++) {
            uint32_t i = z * (subdivisions + 1) + x;
            m.add_face({i, i + (uint32_t)(subdivisions + 1), i + (uint32_t)(subdivisions + 2), i + 1});
        }
    }
    m.compute_normals();
    return m;
}

Mesh create_arrow(float length) {
    Mesh m; m.name = "Arrow";
    Mesh shaft = create_cylinder(0.05f, length * 0.7f, 12);
    for (auto& v : shaft.vertices) v.position.y += length * 0.15f;
    Mesh head = create_cone(0.2f, length * 0.3f, 12);
    for (auto& v : head.vertices) v.position.y += length * 0.5f + length * 0.15f;
    uint32_t offset = (uint32_t)m.vertices.size();
    for (auto& v : shaft.vertices) m.add_vertex(v.position, v.normal, v.uv, v.color);
    for (auto& f : shaft.faces) {
        std::vector<uint32_t> fi;
        for (auto& vi : f.vertices) fi.push_back(vi + offset);
        m.add_face(fi);
    }
    offset = (uint32_t)m.vertices.size();
    for (auto& v : head.vertices) m.add_vertex(v.position, v.normal, v.uv, v.color);
    for (auto& f : head.faces) {
        std::vector<uint32_t> fi;
        for (auto& vi : f.vertices) fi.push_back(vi + offset);
        m.add_face(fi);
    }
    m.compute_normals();
    return m;
}

Mesh create_torus_knot(int p, int q, float major_r, float tube_r) {
    Mesh m; m.name = "TorusKnot";
    int segments = 128, tube_seg = 16;
    for (int i = 0; i <= segments; i++) {
        float t = 2.0f * (float)M_PI * i / segments;
        float r = major_r * (2.0f + cosf((float)q * t)) / 3.0f;
        vec3 center(r * cosf((float)p * t), r * sinf((float)p * t), major_r * sinf((float)q * t) / 3.0f);

        float t2 = t + 0.01f;
        float r2 = major_r * (2.0f + cosf((float)q * t2)) / 3.0f;
        vec3 next(r2 * cosf((float)p * t2), r2 * sinf((float)p * t2), major_r * sinf((float)q * t2) / 3.0f);
        vec3 tangent = glm::normalize(next - center);

        vec3 up = vec3(0, 1, 0);
        if (fabsf(glm::dot(tangent, up)) > 0.99f) up = vec3(1, 0, 0);
        vec3 side = glm::normalize(glm::cross(tangent, up));
        vec3 binormal = glm::cross(tangent, side);

        for (int j = 0; j <= tube_seg; j++) {
            float phi = 2.0f * (float)M_PI * j / tube_seg;
            vec3 pos = center + tube_r * (cosf(phi) * side + sinf(phi) * binormal);
            vec3 nrm = glm::normalize(pos - center);
            vec2 uv((float)i / segments, (float)j / tube_seg);
            m.add_vertex(pos, nrm, uv);
        }
    }
    for (int i = 0; i < segments; i++) {
        for (int j = 0; j < tube_seg; j++) {
            uint32_t a = i * (tube_seg + 1) + j;
            uint32_t b = (i + 1) * (tube_seg + 1) + j;
            m.add_face({a, b, b + 1, a + 1});
        }
    }
    m.compute_normals();
    return m;
}

Mesh create_ico_sphere(float radius, int subdivisions) {
    Mesh m; m.name = "IcoSphere";
    float t = (1.0f + sqrtf(5.0f)) / 2.0f;
    float len = sqrtf(1 + t * t);
    vec3 ico_verts[12] = {
        glm::normalize(vec3(-1, t, 0)) * radius, glm::normalize(vec3(1, t, 0)) * radius,
        glm::normalize(vec3(-1, -t, 0)) * radius, glm::normalize(vec3(1, -t, 0)) * radius,
        glm::normalize(vec3(0, -1, t)) * radius, glm::normalize(vec3(0, 1, t)) * radius,
        glm::normalize(vec3(0, -1, -t)) * radius, glm::normalize(vec3(0, 1, -t)) * radius,
        glm::normalize(vec3(t, 0, -1)) * radius, glm::normalize(vec3(t, 0, 1)) * radius,
        glm::normalize(vec3(-t, 0, -1)) * radius, glm::normalize(vec3(-t, 0, 1)) * radius
    };
    for (auto& v : ico_verts) m.add_vertex(v, glm::normalize(v));

    uint32_t faces[20][3] = {
        {0,11,5},{0,5,1},{0,1,7},{0,7,10},{0,10,11},
        {1,5,9},{5,11,4},{11,10,2},{10,7,6},{7,1,8},
        {3,9,4},{3,4,2},{3,2,6},{3,6,8},{3,8,9},
        {4,9,5},{2,4,11},{6,2,10},{8,6,7},{9,8,1}
    };
    for (auto& f : faces) m.add_face({f[0], f[1], f[2]});

    for (int s = 0; s < subdivisions; s++) m.subdivide();
    for (auto& v : m.vertices) {
        v.normal = glm::normalize(v.position);
        v.position = v.normal * radius;
    }
    m.dirty = true;
    m.compute_normals();
    return m;
}

const std::map<std::string, PrimitiveCreator>& get_primitive_creators() {
    static std::map<std::string, PrimitiveCreator> creators = {
        {"cube", [](float s){ return create_cube(s); }},
        {"sphere", [](float s){ return create_sphere(s * 0.5f, 64, 32); }},
        {"cylinder", [](float s){ return create_cylinder(s * 0.5f, s, 48); }},
        {"cone", [](float s){ return create_cone(s * 0.5f, s, 48); }},
        {"torus", [](float s){ return create_torus(s * 0.4f, s * 0.15f, 48, 24); }},
        {"plane", [](float s){ return create_plane(s * 2.0f); }},
        {"torus_knot", [](float s){ return create_torus_knot(2, 3, s * 0.5f, s * 0.08f); }},
        {"ico_sphere", [](float s){ return create_ico_sphere(s * 0.5f); }},
        {"arrow", [](float s){ return create_arrow(s); }}
    };
    return creators;
}

} // namespace ocp
